aws/resource_aws_ram_resource_share_accepter.go

```diff
@@ -0,0 +1,297 @@
+		},
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(5 * time.Minute),
+			Delete: schema.DefaultTimeout(5 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"invitation_arn": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				Computed:     true,
+				ValidateFunc: validateArn,
+			},
+
+			"share_arn": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				Computed:     true,
+				ValidateFunc: validateArn,
+			},
+
+			"share_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"status": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"receiver_account_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"sender_account_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"share_name": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"resources": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+		},
+	}
+}
+
+func resourceAwsRamResourceShareAccepterCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ramconn
+
+	in := &ram.AcceptResourceShareInvitationInput{
+		ClientToken: aws.String(resource.UniqueId()),
+	}
+
+	if v, ok := d.GetOk("invitation_arn"); ok {
+		in.ResourceShareInvitationArn = aws.String(v.(string))
+	} else if v, ok := d.GetOk("share_arn"); ok {
+		// need to find invitation arn
+		invitationARN, err := resourceAwsRamResourceShareGetInvitationARN(conn, v.(string))
+		if err != nil {
+			return err
+		}
+
+		in.ResourceShareInvitationArn = aws.String(invitationARN)
+	} else {
+		return fmt.Errorf("Either an invitation ARN or share ARN are required")
+	}
+
+	log.Printf("[DEBUG] Accept RAM resource share invitation request: %s", in)
+	out, err := conn.AcceptResourceShareInvitation(in)
+	if err != nil {
+		return fmt.Errorf("Error accepting RAM resource share invitation: %s", err)
+	}
+
+	d.SetId(aws.StringValue(out.ResourceShareInvitation.ResourceShareInvitationArn))
+
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{ram.ResourceShareInvitationStatusPending},
+		Target:  []string{ram.ResourceShareInvitationStatusAccepted},
+		Refresh: resourceAwsRamResourceShareAccepterStateRefreshFunc(conn, d.Id()),
+		Timeout: d.Timeout(schema.TimeoutCreate),
+	}
+
+	_, err = stateConf.WaitForState()
+	if err != nil {
+		return fmt.Errorf("Error waiting for RAM resource share (%s) state: %s", d.Id(), err)
+	}
+
+	return resourceAwsRamResourceShareAccepterRead(d, meta)
+}
+
+func resourceAwsRamResourceShareAccepterRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ramconn
+
+	request := &ram.GetResourceShareInvitationsInput{
+		ResourceShareInvitationArns: []*string{aws.String(d.Id())},
+	}
+
+	out, err := conn.GetResourceShareInvitations(request)
```

Comment:
Given the requested ID change above, we should lookup by `ResourceShareArns` instead of `ResourceShareInvitationArns`. This will allow this resource to tolerate the invitation being rescinded and the resource share reaccepted under a new ARN outside of Terraform.

We should use the paginated AWS Go SDK function [`GetResourceShareInvitationsPages`](https://docs.aws.amazon.com/sdk-for-go/api/service/ram/#RAM.GetResourceShareInvitationsPages) here, to prevent any potential issues with paging. 

```go
input := &ram.GetResourceShareInvitationsInput{
	ResourceShareArns: []*string{aws.String(d.Id())},
}

var invitation *ram.ResourceShareInvitation
err := conn.GetResourceShareInvitationsPages(input, func(page *ram.GetResourceShareInvitationsOutput, lastPage bool) bool {
	for _, rsi := range page.ResourceShareInvitations {
		if aws.StringValue(rsi.Status) == ram.ResourceShareInvitationStatusAccepted {
			invitation = rsi
			break
		}
	}

	return !lastPage
})

if isAWSErr(err, ram.ErrCodeUnknownResourceException, "") {
	log.Printf("[WARN] No RAM resource share invitation by ARN (%s) found, removing from state", d.Id())
	d.SetId("")
	return nil
}

if err != nil {
	return fmt.Errorf("Error reading RAM resource share invitation %s: %s", d.Id(), err)
}

if invitation == nil {
	log.Printf("[WARN] No RAM resource share invitation by ARN (%s) found, removing from state", d.Id())
	d.SetId("")
	return nil
}
```

Reply:
I'm concerned that the issues with the invitation ARN are going to cause problems. If the receiver account has multiple resource shares with the sender account, the read function is going to randomly grab the first invitation AWS returns with the accepted status. When deleting, it can leave the wrong share.

Reply:
The API filtering should be limiting to a single resource share, but in case its not, we can be extra cautious with also checking the resource share ARN matches our ID:

```go
var invitation *ram.ResourceShareInvitation
err := conn.GetResourceShareInvitationsPages(input, func(page *ram.GetResourceShareInvitationsOutput, lastPage bool) bool {
	for _, rsi := range page.ResourceShareInvitations {
		if aws.StringValue(rsi.ResourceShareArn) == d.Id() && aws.StringValue(rsi.Status) == ram.ResourceShareInvitationStatusAccepted {
			invitation = rsi
			return false
		}
	}

	return !lastPage
})
```

Reply:
That extra check doesn't address the concern: you can have multiple invitations for the same share. **But**, after testing it doesn't appear to matter which invitation you accept. Accepting any one of them has the same effect as accepting any other of them. I'm good if you're good with this.

This is the config I used to test it out. Here we don't know which invitation will be accepted since `aws_ram_principal_association.test2.resource_share_arn == aws_ram_principal_association.test1.resource_share_arn`, but, it doesn't appear to matter.

```hcl
resource "aws_ram_principal_association" "test1" {
  provider           = "aws.alternate"
  principal          = "${data.aws_caller_identity.receiver.account_id}"
  resource_share_arn = "${aws_ram_resource_share.test.arn}"
}

resource "aws_ram_principal_association" "test2" {
  provider           = "aws.alternate"
  principal          = "${data.aws_caller_identity.receiver.account_id}"
  resource_share_arn = "${aws_ram_resource_share.test.arn}"
}

data "aws_caller_identity" "receiver" {}

resource "aws_ram_resource_share_accepter" "test" {
  share_arn = "${aws_ram_principal_association.test2.resource_share_arn}"
}
```